### PR TITLE
fix: nothing to mutate now costs nothing

### LINKED ISSUE #76: fix: --daemon boots the app N+1 times when the job list is empty
Found by the adversarial reviewer during the #58 review. Pre-existing, not introduced by that refactor.

## What
`mutineer run ... --rails --daemon --since origin/main --jobs 4` on a PR that changes no mutatable source line boots the app **five times** and scores nothing.

## Why it matters
README documents exactly this invocation for PR CI (`since: origin/${{ github.base_ref }}`), so a docs-only or test-only PR is the normal case, not an edge case. The verdicts are not wrong (there are none), so this is wasted time and astonishment rather than a score lie, but on a real Rails app it is five full boots for zero work.

## The path
1. `Runner.filter_since` returns an empty `jobs` array.
2. `DaemonBackend.build_coverage_map` still runs, booting a daemon and capturing full-suite coverage for zero mutants.
3. The worker cap `worker_count = [worker_count, jobs.size].min if jobs.size.positive?` is skipped by its own `positive?` guard, so `worker_count` stays at the user`s `--jobs 4`.
4. `run_parallel` spawns four more daemons, whose threads immediately hit `ThreadError` on the empty queue and quit.

## How
Return early in `DaemonBackend.execute`, right after the `filter_since` line and before `build_coverage_map`:

```ruby
return [AggregateResult.new(ignored_results), source_map] if jobs.empty?
```

That is the same shape the method already returns at its end, and it skips both the coverage daemon and every worker daemon.

## Acceptance criteria
- [ ] An empty job list on the daemon path boots nothing
- [ ] A test pins the boot count for the empty case (none exists today)
- [ ] The in-process path is checked for the same shape

### PR DESCRIPTION
**Pull request by davidteren**
Closes #76. Found by the adversarial reviewer during the #58 gate.

## What

`mutineer run --rails --daemon --since origin/main --jobs 4` on a PR that changes no mutatable line booted the app **five times** to score zero mutants: once for the coverage map, then once per worker. Both backends that can know the job list is empty before paying now return immediately.

## Why it matters

README documents `--since origin/${{ github.base_ref }}` as the PR-CI invocation, so a docs-only or test-only PR is the normal case, not an edge case. On a real Rails app that was minutes of boots to produce an empty report.

The worker cap `[worker_count, jobs.size].min` is skipped by its own `positive?` guard, so an empty job list left `worker_count` at whatever `--jobs` the user passed.

## Scope

- **`--daemon`** — returns before `build_coverage_map`, so nothing boots.
- **`--test-command`** — returns before `ExternalBackend.smoke_check!`, which otherwise runs the whole `--test` set to calibrate a per-mutant timeout no mutant would use. The predictability lens pointed out the waste was never daemon-specific, and leaving it would have made the two backends answer the same input differently.
- **in-process** — unchanged, and cannot do this. It boots and builds coverage *before* `collect_jobs`, so at that point it does not yet know the list is empty. Reordering would fight the constraint that `Coverage.start` must precede the boot require.

## The trade-off, decided deliberately

A zero-job run used to reach `DaemonClient.start` (or the smoke check), so an app that could not boot, an invalid `--test-command`, or a red unmutated suite failed that run with exit 1. Nothing runs now, so it exits 0.

Cubic flagged this at P1 on both backends, as masking a real signal. The predictability lens flagged the same thing at P3 and called the new behaviour the defensible one. It is a product question rather than a bug, so it went to the maintainer, who chose to skip: Mutineer was asked about changed lines, there are none, and the repo's own test job already covers whether the app boots. It is stated in the CHANGELOG rather than left for someone to discover.

## Testing

Two tests, each verified to fail when its guard is removed:

- `test_an_empty_job_list_boots_nothing` — flunks if any `DaemonClient` is constructed
- `test_an_empty_job_list_skips_the_external_smoke_check` — flunks if the smoke check runs

Also checked by hand: the early return skips no cleanup (`FileSwap.restore_orphans` already ran before `collect_jobs`, and nothing has been mutated), and the returned aggregate is identical to what the full path builds for zero jobs, since `results` is empty there.

`rake test` 420 runs / 1203 assertions, `rake test:daemon` 14 runs, `yard:strict` 100%.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01EfCKGYtEpcYLtp71Rsqf3h

---

<a href="https://app.devin.ai/review/davidteren/mutineer/pull/81" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

---
## Summary by cubic
Return early when there are no mutation jobs so docs-only or test-only PRs finish instantly. Also sweep orphaned daemon temp files so empty runs don’t break app boot.

- **Bug Fixes**
  - `--daemon`: return before building the coverage map or starting workers when the job list is empty.
  - Sweep orphaned `mutineer_daemon*.rb` before returning to prevent Zeitwerk boot errors on empty runs.
  - External `--test-command`: return before `ExternalBackend.smoke_check!` if the job list is empty.
  - Drop the dead worker-cap guard (`jobs.size.positive?`) since the early return guarantees a non-empty list before capping.
  - In-process path unchanged; it must boot before collecting jobs.

- **Migration**
  - Zero-job `--daemon` runs now exit 0 (no work) instead of 1 if the app cannot boot.

<sup>Written for commit 0c86d805428b0d6f9e1006b1d80e64d998692aca. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/davidteren/mutineer/pull/81?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: lib/mutineer/daemon_backend.rb [hunk 1]
```diff
@@ -44,6 +44,12 @@ def self.execute(config, operator_classes)
       jobs = Runner.filter_since(jobs, source_map, config) if config.since
       abs_tests = config.tests.map { |t| File.expand_path(t, config.project_root) }
 
+      # Nothing to mutate (`--since` matched no changed line, or every mutant is
+      # suppressed). Return before booting anything: the coverage daemon and the
+      # worker daemons below each boot the whole app, and README documents
+      # `--since origin/<base>` for PR CI, where a docs-only PR is routine.
+      return [AggregateResult.new(ignored_results), source_map] if jobs.empty?
```
[the remark this answers is not in the document]
Comment (davidteren) [reply lib/mutineer/daemon_backend.rb]: **Claude:** Valid, and it was a regression I introduced. Fixed in `9294e64`.

Confirmed the mechanism before fixing: `DaemonServer#sweep_temps` clears `mutineer_daemon*.rb` from the source dirs at boot (`daemon_server.rb:112`), and the early return means nothing boots, so nothing sweeps. Worse than a wasted boot, as you imply — the temp sits in a directory like `app/models`, so Zeitwerk tries to autoload it and the **app's own** boot breaks, not just the next Mutineer run.

`DaemonBackend.execute` now sweeps tool-side before returning. `Runner.sweep_orphans` takes the glob so the daemon pattern reuses it instead of growing a second copy of the same three lines. Test added and verified to fail without the sweep.

The external path needed no equivalent: `FileSwap.restore_orphans` already runs before `collect_jobs`, ahead of its early return.

Worth recording that this is the first review from the cubic GitHub app on this repo, and it caught something the local `cubic review -b` run on the same diff did not.

File: lib/mutineer/daemon_backend.rb
[the remark this answers is not in the document]
Comment (davidteren) [reply lib/mutineer/daemon_backend.rb:1]: **Claude:** Good catch — fixed in `0c86d80` rather than left as harmless. You are right that after the early return `jobs.size.positive?` is unconditionally true there, and a dead guard invites the reader to infer that an empty list can still reach that line, which is exactly the wrong mental model now.

File: lib/mutineer/runner.rb [hunk 2]
```diff
@@ -192,6 +192,12 @@ def self.execute_external(config, operator_classes)
       jobs, ignored_results, source_map = collect_jobs(config, operator_classes)
       jobs = filter_since(jobs, source_map, config) if config.since
 
+      # Nothing to mutate: return before the smoke check, which runs the whole
+      # --test set to calibrate a timeout no mutant would use. Mirrors the daemon
+      # path (#76). The in-process path cannot do this — it boots and builds
+      # coverage before collect_jobs, so it does not yet know the list is empty.
+      return [AggregateResult.new(ignored_results), source_map] if jobs.empty?
```
[the remark this answers is not in the document]
Comment (davidteren) [reply lib/mutineer/runner.rb:197]: **Claude:** Thanks for verifying it rather than assuming — that is the property the whole change rests on, and you checked the same two things I did by hand: `results` is empty for zero jobs so the aggregate is identical, and the skipped `ensure FileSwap.restore_orphans` cannot matter because nothing was mutated and the pre-emptive restore already ran before `collect_jobs`.

Worth noting cubic found the one gap in that reasoning on the daemon side: its orphan sweep happens **inside** the daemon at boot, so the early return skipped it. Fixed in `9294e64`.

No change from this thread.

File: lib/mutineer/runner.rb [hunk 3]
```diff
@@ -192,6 +192,12 @@ def self.execute_external(config, operator_classes)
       jobs, ignored_results, source_map = collect_jobs(config, operator_classes)
       jobs = filter_since(jobs, source_map, config) if config.since
 
+      # Nothing to mutate: return before the smoke check, which runs the whole
+      # --test set to calibrate a timeout no mutant would use. Mirrors the daemon
+      # path (#76). The in-process path cannot do this — it boots and builds
+      # coverage before collect_jobs, so it does not yet know the list is empty.
```
[the remark this answers is not in the document]
Comment (davidteren) [reply lib/mutineer/runner.rb]: **Claude:** Trimmed in `0c86d80`. The external comment is now two lines: what is skipped and why. The comparison with the in-process path was the part that read as design history, and it lives in the PR body and the CHANGELOG where a reader looking for rationale will find it.
